kratos/utilities/auxiliar_model_part_utilities.cpp

```diff
@@ -12,14 +12,94 @@
 //
 
 // System includes
+#include <unordered_set>
 
 // External includes
 
 // Project includes
+#include "includes/key_hash.h"
 #include "utilities/auxiliar_model_part_utilities.h"
 
 namespace Kratos
 {
+void AuxiliarModelPartUtilities::RecursiveEnsureModelPartOwnsProperties(const bool RemovePreviousProperties)
+{
+    // First we do in this model part
+    EnsureModelPartOwnsProperties(RemovePreviousProperties);
+
+    // Now we do in submodelparts
+    for (auto& r_sub_model_part : mrModelPart.SubModelParts()) {
+        AuxiliarModelPartUtilities(r_sub_model_part).RecursiveEnsureModelPartOwnsProperties(RemovePreviousProperties);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void AuxiliarModelPartUtilities::EnsureModelPartOwnsProperties(const bool RemovePreviousProperties)
+{
+    // First we clear the properties if we want so
+    if (RemovePreviousProperties) {
+        mrModelPart.GetMesh(0).pProperties()->clear();
+    }
+
+    // The list of properties
+    std::unordered_set<Properties::Pointer, IndexedObjecPointertHasher<Properties::Pointer>, IndexedObjectPointerComparator<Properties::Pointer>> list_of_properties;
+
+    // Iterating over the elements
+    auto& r_elements_array = mrModelPart.Elements();
+    const auto it_elem_begin= r_elements_array.begin();
+    const int number_of_elements = static_cast<int>(r_elements_array.size());
+
+    // Iterating over the conditions
+    auto& r_conditions_array = mrModelPart.Conditions();
+    const auto it_cond_begin= r_conditions_array.begin();
+    const int number_of_conditions = static_cast<int>(r_conditions_array.size());
+
+    #pragma omp parallel
+    {
+        // The list of properties
+        std::unordered_set<Properties::Pointer, IndexedObjecPointertHasher<Properties::Pointer>, IndexedObjectPointerComparator<Properties::Pointer>> buffer_list_of_properties;
+
+        #pragma omp for schedule(guided, 512) nowait
```

Comment:
why guided and not static (or even dynamic)?

Reply:
Certainly I think is copy paste from other side, I can change it